### DISCUSSION TITLE
MailKit.Security.AuthenticationException : 'Authentication failed.'

### DISCUSSION THREAD

**Discussion by fjquen** [jstedfast/MailKit#1680] [Q&A] [answered]
Hello, sorry to bother you, but I have a problem with MailKit. I'm getting a MailKit.Security.AuthenticationException error: 'Authentication failed.' It occurs on the line where there is await client.AuthenticateAsync(oauth2);. Note that in oauth2, there is var oauth2 = new SaslMechanismOAuth2("*******@outlook.com", authToken.AccessToken);. So, should I remove the string or find another solution? Thanks for following along so far and I used the code from this link: https://github.com/jstedfast/MailKit/blob/master/ExchangeOAuth2.md#authenticating-a-web-service-with-oauth2.

**Comment by jstedfast** [comment]
As far as I know, there's no reason you can't use a hard-coded string for the username.

I don't know why authentication is failing for you. If you followed all of the steps, it should be working.

Unfortunately, this is not an issue that I'm equipped to help you debug if it doesn't work. You might have to reach out to the Microsoft  Office365 team to help diagnose what the issue is. They'd likely be able to track telemetry log data to pinpoint what went wrong.

**Comment by fjquen** [comment]
Thank you for the response. It's nice of you to take the time to reply. I will try to contact Microsoft. For now, the only code I've followed is this :

```
class Program
{
    public static async Task Main(string[] args)
    {
        string clientSecret = "******";
        string clientId = "*********";
        string tenantId = "*********";
        var confidentialClientApplication = ConfidentialClientApplicationBuilder.Create(clientId)
        .WithAuthority($"https://login.microsoftonline.com/{tenantId}/v2.0")
        .WithClientSecret (clientSecret)
        .Build();

        var scopes = new string[] {
            // For IMAP and POP3, use the following scope
            "https://outlook.office365.com/.default"
        };

        var authToken = await confidentialClientApplication.AcquireTokenForClient(scopes).ExecuteAsync();
        var oauth2 = new SaslMechanismOAuth2("*****@outlook.com", authToken.AccessToken);
        using (var client = new ImapClient(new ProtocolLogger("imap_mail_move.log")))
        {
            try
            {
                await client.ConnectAsync("outlook.office365.com", 993, SecureSocketOptions.SslOnConnect);
                await client.AuthenticateAsync(oauth2);
                await client.DisconnectAsync(true);
            }
            catch (AuthenticationException ex)
            {
                Console.WriteLine($"Authentication error: {ex.Message}");
                Console.WriteLine($"Error details: {ex.InnerException?.Message}");
            }
        }
    }
}
```


Would I have by chance made a mistake in the above code?

**Reply by jstedfast** [reply]
The scope is different in my guide:

```csharp
var scopes = new string[] {
    // For IMAP and POP3, use the following scope
    "https://ps.outlook.com/.default"

    // For SMTP, use the following scope
    // "https://outlook.office365.com/.default"
};
```

I can't be sure that that matters, but it might be worth checking.

**Reply by fjquen** [reply]
Ok, thanks. I have this result in my IMAP log file. Have you seen this before?
Connected to imaps://outlook.office365.com:993/
S: * OK The Microsoft Exchange IMAP4 service is ready. [UABBADcAUAAyADYANABDAEEAMAA0ADUAMQAuAEYAUgBBAFAAMgA2ADQALgBQAFIATwBEAC4ATwBVAFQATABPAE8ASwAuAEMATwBNAA==]
C: A00000000 CAPABILITY
S: * CAPABILITY IMAP4 IMAP4rev1 AUTH=PLAIN AUTH=XOAUTH2 SASL-IR UIDPLUS ID UNSELECT CHILDREN IDLE NAMESPACE LITERAL+
S: A00000000 OK CAPABILITY completed.
C: A00000001 AUTHENTICATE XOAUTH2 ********
S: A00000001 NO AUTHENTICATE failed.

**Reply by jstedfast** [reply]
That's what every failed XOAUTH2 authentication attempt looks like :-)

**Reply by fjquen** [reply]
ok thank Can I dismantle xoauth2 or see with an administrator how to set up basic authentication ?

**Reply by jstedfast** [reply]
Your log indicates that the PLAIN mechanism is supported, whether it will work or not is anyone's guess but you could try it.

```csharp
await client.ConnectAsync("outlook.office365.com", 993, SecureSocketOptions.SslOnConnect);
await client.AuthenticateAsync(new SaslMechanismPlain("username@outlook.com", "password"));
```

**Reply by fjquen** [reply, marked as answer]
I found the solution with the AcquireTokenByUsernamePassword function thank for help @jstedfast
